src/hotspot/share/utilities/debug.cpp

```diff
@@ -671,10 +671,35 @@ extern "C" bool dbg_is_safe(const void* p, intptr_t errvalue) {
   return p != nullptr && SafeFetchN((intptr_t*)const_cast<void*>(p), errvalue) != errvalue;
 }
 
+
 extern "C" bool dbg_is_good_oop(oopDesc* o) {
   return dbg_is_safe(o, -1) && dbg_is_safe(o->klass(), -1) && oopDesc::is_oop(o) && o->klass()->is_klass();
 }
 
+// Additional "good oop" checks, separate method to not disturb existing asserts.
+extern "C" bool dbg_is_good_oop_detailed(oopDesc* o) {
```

Comment:
What was the motivation for adding this? Is this copied from other HS code? How do we know it's doing a good enough job of verification? What happens if verification is not thorough enough?

Reply:
On use of dbg_is_good_oop() and do we need our own dbg_is_good_oop_detailed():

The worst case is a crash if the data looks enough like an oop to pass basic checks, but if the klass pointer or a field turns out to be a wild pointer.  So I wanted a stricter and more careful "is good oop" check, e.g. to make it verify the klass pointer is safe, before calling is_klass().  I cannot now see a crash when taking a valid object and examining thousands of what must be wrong pointers in memory after it.

I kept these separate, as had seen an assert when making the existing dbg_is_good_oop more strict.  But that was an error, I had the pointer arithmetic wrong.

This check is called by asserts in two existing places, plus the new diagnosticCommand in this PR.  With this update I'm running all of tier1 OK, which includes ContFramePopTest.java where I could previously trigger an assert.